// react-context-menu/lib/@components/MenuItem.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MenuItem = function MenuItem(_ref) {
  var item = _ref.item;
  var label = item.label,
      icon = item.icon,
      onClick = item.onClick;


  return _react2.default.createElement(
    'span',
    {
      className: 'menuItem',
      onClick: onClick,
      key: label,
      style: { "cursor": "pointer", "fontSize": "14px", "display": "flex", "alignItems": "center", "justifyContent": "flex-start", "marginBottom": "7px" }
    },
    icon && _react2.default.createElement('img', { className: 'icon', src: icon }),
    label
  );
};

MenuItem.propTypes = process.env.NODE_ENV !== "production" ? {
  item: _propTypes2.default.shape({
    label: _propTypes2.default.string.isRequired,
    onClick: _propTypes2.default.func.isRequired,
    icon: _propTypes2.default.string
  }).isRequired
} : {};

exports.default = MenuItem;
module.exports = exports['default'];
